### PR TITLE
[SCB-359]Servicecomb support latency statistic

### PR DESCRIPTION
**Pull request by heyile**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by yhs0092** [comment]
Hi, maybe you can merge these three commit into one by `git rebase` : )

And I don't quite understand the meaning of your first commit, could you tell me?

**Comment by heyile** [comment]
> Hi, maybe you can merge these three commit into one by `git rebase` : )
> 
> And I don't quite understand the meaning of your first commit, could you tell me?
the first commit is because that the name of variable may mislead others... and I want distinguish it from latency statistic. maybe `rebase` last two commits is better :-)

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/PublishUtils.java [hunk 1]
```diff
@@ -33,7 +34,12 @@ private PublishUtils() {
 
   public static PerfInfo createPerfInfo(MeasurementNode stageNode) {
     PerfInfo perfInfo = new PerfInfo();
-
+    MeasurementNode timeInterval = stageNode.findChild(StatisticsBucket.TIME_INTERVAL);
+    // latency buckets
+    if (timeInterval != null) {
+      perfInfo.initInterval(timeInterval);
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/PublishUtils.java]: latency bucket is not belongs to perfInfo, right?

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  public void calcMeasurements(long msNow, long secondInterval) {
+    List<Measurement> measurements = new ArrayList<>();
+    calcMeasurements(measurements, msNow, secondInterval);
+    allMeasurements = measurements;
+  }
+
+  @Override
+  public void calcMeasurements(List<Measurement> measurements, long msNow, long secondInterval) {
+    latencyIntervals.forEach(latencyInterval -> {
+      measurements.add(newMeasurement(rootId.withTag("buckets_interval", latencyInterval.toString()),
```
[the remark this answers is not in the document]
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: ok. I see

[the remark this answers is not in the document]
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: done

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    int[] arrays = Arrays.stream(property.trim().split(","))
+        .mapToInt(Integer::parseInt)
+        .distinct()
+        .sorted()
+        .filter(num -> num > 0)
+        .toArray();
+    if (arrays == null || arrays.length == 0) {
+      return results;
+    }
+    results.add(new LatencyInterval(0L, arrays[0] * MS_UNIT));
```
[the remark this answers is not in the document]
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: done

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    for (int i = 1; i < arrays.length; i++) {
+      results.add(new LatencyInterval(arrays[i - 1] * MS_UNIT, arrays[i] * MS_UNIT));
+    }
+    results.add(new LatencyInterval(arrays[arrays.length - 1] * MS_UNIT, Long.MAX_VALUE));
+    return results;
+  }
+
+  public static String getStrFormatFromProperty(List<LatencyInterval> latencyIntervals) {
```
[the remark this answers is not in the document]
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: I have moved it to the Class **DefaultLogPublisher**

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    public boolean match(long num) {
+      return minValue <= num && maxValue > num;
+    }
+
+    public void increment() {
+      this.times.increment();
+    }
+
+    public long getAndUpdateLastTimes() {
+      long result = times.longValue() - lastUpdated;
```
[the remark this answers is not in the document]
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: done

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java [hunk 6]
```diff
@@ -43,18 +44,26 @@
...
 
+    totalBucket = creatStageBucket(MeterInvocationConst.TOTAL_BUCKETS);
     totalTimer = creatStageTimer(MeterInvocationConst.STAGE_TOTAL);
     prepareTimer = creatStageTimer(MeterInvocationConst.STAGE_PREPARE);
     handlersRequestTimer = creatStageTimer(MeterInvocationConst.STAGE_HANDLERS_REQUEST);
     handlersResponseTimer = creatStageTimer(MeterInvocationConst.STAGE_HANDLERS_RESPONSE);
   }
 
+  protected StatisticsBucket creatStageBucket(String bucketsValue) {
+    return new StatisticsBucket(id.withTag(MeterInvocationConst.TAG_STAGE, bucketsValue));
```
[the remark this answers is not in the document]
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java]: So.  what value should I use here ?

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java [hunk 7]
```diff
@@ -60,27 +62,31 @@
 
   //sample
   private static final String SIMPLE_HEADER = "%s:\n  simple:\n"
-      + "    status          tps      latency             operation\n";
+      + "    status      tps    latency            %soperation\n";
```
[the remark this answers is not in the document]
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: done.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java [hunk 8]
```diff
@@ -89,7 +95,9 @@ public void init(GlobalRegistry globalRegistry, EventBus eventBus, MetricsBootst
         .get()) {
       return;
     }
-
+    latencyIntervals = StatisticsBucket
```
[the remark this answers is not in the document]
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: ... maybe it's doesn't matter ? Must I leave spaces here ?

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/PublishUtils.java (hunk 1 above)
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/PublishUtils.java]: Yes. I move it to the class **OperationPerf**, as I thought that every operation have it's own latency 
```java
public class OperationPerf {
  private String operation;

  private Map<String, PerfInfo> stages = new HashMap<>();

  private Map<String, BucketInfo> buckets = new HashMap<>();
```

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: maybe **buckets_interval** or **buckets_latency** is better ?

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    private long lastUpdated = 0L;
+
+    private long minValue;
+
+    private long maxValue;
+
+    public LatencyInterval(long minValue, long maxValue, Id rootId) {
+      this.minValue = minValue;
+      this.maxValue = maxValue;
+      if (rootId != null) {
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: when will rootId be null?

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+  public static final String METRICS_LATENCY = DynamicPropertyFactory.getInstance()
+      .getStringProperty("servicecomb.metrics.invocation.latency.buckets", "0")
+      .get();
+
+  private Id rootId;
+
+  private List<LatencyInterval> latencyIntervals;
+
+  public StatisticsBucket(Id id) {
+    this.id = id;
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: why need both id and rootId?

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+      int[] arrays = Arrays.stream(property.trim().split(","))
+          .mapToInt(Integer::parseInt)
+          .distinct()
+          .sorted()
+          .filter(num -> num > 0)
+          .toArray();
+      if (arrays == null || arrays.length == 0) {
+        return results;
+      }
+      results.add(new LatencyInterval(0L, arrays[0] * TimeUnit.MILLISECONDS.toNanos(1), rootId));
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: TimeUnit.MILLISECONDS.toNanos(arrays[0])

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    public long getAndUpdateLastTimes() {
+      long currentTimes = times.longValue();
+      long result = currentTimes - lastUpdated;
+      this.lastUpdated = currentTimes;
+      return result;
+    }
+
+    public String getStandardLatencyStr() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("[").append(minValue / TimeUnit.MILLISECONDS.toNanos(1))
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/StatisticsBucket.java]: TimeUnit.NANOSECONDS.toMillis(minValue)

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+import java.util.concurrent.atomic.LongAdder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.config.DynamicPropertyFactory;
+import com.netflix.spectator.api.Id;
+import com.netflix.spectator.api.Measurement;
+
+public class BucketsDistribution extends AbstractPeriodMeter {
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: LatencyDistribution

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        results.add(new LatencyInterval(arrays[i - 1], arrays[i], rootId));
+      }
+      results.add(new LatencyInterval(arrays[arrays.length - 1], Long.MAX_VALUE, rootId));
+    } catch (Throwable e) {
+      LOGGER.error("Failed to parse property servicecomb.metrics.invocation.latency.buckets", e);
+    }
+    return results;
+  }
+
+  public static class LatencyInterval {
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: LatencyScope

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+import org.slf4j.LoggerFactory;
+
+import com.netflix.config.DynamicPropertyFactory;
+import com.netflix.spectator.api.Id;
+import com.netflix.spectator.api.Measurement;
+
+public class BucketsDistribution extends AbstractPeriodMeter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BucketsDistribution.class);
+
+  public static final String METRICS_LATENCY = DynamicPropertyFactory.getInstance()
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: default to ""

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+
+    private long lastUpdated = 0L;
+
+    private long minValue;
+
+    private long maxValue;
+
+    public LatencyInterval(long minValue, long maxValue, Id rootId) {
+      this.minValue = TimeUnit.MILLISECONDS.toNanos(minValue);
+      this.maxValue = maxValue == Long.MAX_VALUE ? Long.MAX_VALUE : TimeUnit.MILLISECONDS.toNanos(maxValue);
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: MAX-VALUE -1 will overflow also.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java [hunk 17]
```diff
@@ -43,20 +44,30 @@
   // handler response
   private SimpleTimer handlersResponseTimer;
 
+  // bucket total
+  private BucketsDistribution totalBucket;
+  //stage
+  private Id stageId;
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java]: no need to cache stageId

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java [hunk 18]
```diff
@@ -43,20 +44,30 @@
   // handler response
   private SimpleTimer handlersResponseTimer;
 
+  // bucket total
+  private BucketsDistribution totalBucket;
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java]: change name

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java [hunk 19]
```diff
@@ -58,29 +63,32 @@
...
-      + "        cFiltersReq : %-19s sendReq     : %-19s getConnect  : %-19s writeBuf    : %s\n"
-      + "        waitResp    : %-19s wakeConsumer: %-19s cFiltersResp: %-19s handlersResp: %s\n"
-      + "        sFiltersResp: %-19s sendResp    : %s\n";
+      + "        prepare     : %-18s queue       : %-18s sFiltersReq : %-18s handlersReq : %s\n"
+      + "        cFiltersReq : %-18s sendReq     : %-18s getConnect  : %-18s writeBuf    : %s\n"
+      + "        waitResp    : %-18s wakeConsumer: %-18s cFiltersResp: %-18s handlersResp: %s\n"
+      + "        sFiltersResp: %-18s sendResp    : %s\n";
+
+
+  private List<String> sortedLatencyIntervals;
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: comments these fields with sample configuration

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+      results.add(new LatencyInterval(arrays[arrays.length - 1], Long.MAX_VALUE, rootId));
+    } catch (Throwable e) {
+      LOGGER.error("Failed to parse property servicecomb.metrics.invocation.latency.buckets", e);
+    }
+    return results;
+  }
+
+  public static class LatencyInterval {
+
+    public static final int leastLength = 7;
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: only for publish, not relate to calc......

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+    private long lastUpdated = 0L;
+
+    private long minValue;
+
+    private long maxValue;
+
+    public LatencyInterval(long minValue, long maxValue, Id rootId) {
+      this.minValue = TimeUnit.MILLISECONDS.toNanos(minValue);
+      this.maxValue = maxValue == Long.MAX_VALUE ? Long.MAX_VALUE : TimeUnit.MILLISECONDS.toNanos(maxValue);
+      this.id = rootId.withTag("statistic", getStandardLatencyStr(minValue, maxValue, leastLength));
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: tag no need to same to output format.....
tag is "[0,1)", and output format is "[0,1)  ", seems not conflict, right?

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java [hunk 22]
```diff
@@ -266,34 +294,65 @@ protected void printProducerLog(DefaultPublishModel model, StringBuilder sb) {
...
+    if (operationLatencyDistribution == null || sortedLatencyIntervals == null) {
+      return "";
+    }
+    Object[] array = sortedLatencyIntervals.stream()
+        .map(value -> operationLatencyDistribution.getInterval().getOrDefault(value, 0))
+        .toArray();
+    return String.format(intervalFormat, array);
+  }
+
+  public String getLogStrFormat(List<String> latencyIntervals) {
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: 1.private
2.change name relate to latency distribution

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/model/invocation/OperationLatencyDistribution.java [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+package org.apache.servicecomb.metrics.core.publish.model.invocation;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.servicecomb.foundation.metrics.publish.spectator.MeasurementNode;
+
+import com.google.common.annotations.VisibleForTesting;
+
+public class OperationLatencyDistribution {
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/model/invocation/OperationLatencyDistribution.java]: try to delete this class, make logic simpler

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java (hunk 13 above)
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: Maybe **LatencyDistributionMeter** is better. 
as it's the child of **AbstractPeriodMeter**  and in the package named **meter**

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java (hunk 14 above)
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: Ok, I will change the class name

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java (hunk 15 above)
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: ok, change the default value is ""

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java (hunk 17 above)
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java]: I will removed it

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java (hunk 18 above)
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/meter/invocation/AbstractInvocationMeter.java]: I changed it to **latencyDistribution**

```java
  // latency distribution
  private LatencyDistributionMeter latencyDistribution;
```

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java (hunk 21 above)
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: I think so, I will redesign the method

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java (hunk 22 above)
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: done. the new method is as bellow 
```java
  /**
   * <p>
   *     get respond str format from list of str. for example
   *    {@code
   *         List<String> strList = Arrays.asList("[0,1) ", "[1,4) ", "[4, ) ");
   *         String strFormat = getLatencyScopeFormat(strList);
   *         System.out.println(String.format(strFormat,strList.toArray()));
   *         System.out.println(String.format(strFormat, 1, 23, 433));
   *         System.out.println(String.format(strFormat, 121, 2, 4));
   *    }
   *    In the example above, result is showed as bellow:
   *        [0,1)  [1,4)  [4, )
   *        1      23     433
   *        121    2      4
   * </p>
   * @param latencyIntervals list of str
   * @return str format
   */
  private String getLatencyScopeFormat(List<String> latencyIntervals) {
```

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/model/invocation/OperationLatencyDistribution.java (hunk 23 above)
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/model/invocation/OperationLatencyDistribution.java]: I have replace the class with
```java
public class OperationPerf {
  private String operation;

  private Map<String, PerfInfo> stages = new HashMap<>();

  private List<Integer> latencyDistribution = new ArrayList<>();
```

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java (hunk 21 above)
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]:
```java
  private static Id createLatencyScopeId(String leftStr, String rightStr, Id rootId) {
    return rootId.withTag("statistic", LatencyScope.getStandardLatencyScopeStr(leftStr, rightStr, 0));
  }

    public static String getStandardLatencyScopeStr(String leftStr, String rightStr, int minLength) {
      StringBuilder sb = new StringBuilder();
      sb.append("[").append(leftStr)
          .append(",")
          .append(rightStr)
          .append(")");
      if (sb.length() < minLength) {
        for (int i = 0; i < minLength - sb.length(); i++) {
          sb.append(' ');
        }
      }
      return sb.toString();
    }

```

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java (hunk 20 above)
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: I have moved the **leastLength**  to the class **DefaultLogPublisher**.
as I thought that this number is relate to the log and has no relation with tag

```java
public class DefaultLogPublisher implements MetricsInitializer {
  private static final int LEAST_LATENCY_SCOPE_STR_LENGTH = 7;
```

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java (hunk 13 above)
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/meter/BucketsDistribution.java]: done

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java [hunk 24]
```diff
@@ -266,34 +301,80 @@ protected void printProducerLog(DefaultPublishModel model, StringBuilder sb) {
...
+   *    }
+   *    In the example above, result is showed as bellow:
+   *        [0,1)  [1,4)  [4, )
+   *        1      23     433
+   *        121    2      4
+   * </p>
+   * @param latencyIntervals list of str
+   * @return str format
+   */
+  private String getLatencyScopeFormat(List<String> latencyIntervals) {
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: change parameter name

Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: done

**Pull request merged**